### DISCUSSION TITLE
Source code modifications in MPASO : need an integer for the day of year

### DISCUSSION THREAD

**Discussion by jacob-stu-allen** [E3SM-Project/E3SM#6064] [E3SM model help]
Hello!

Version: E3SMv2_1
Case: WCYCL1850
Res: ne30pg2_EC30to60E2r2

For my source code modifications which override some wind-anomalies to the daily climatology in MPASO, I need to use an integer to subset my array (365 days, ncells) which indicates the day of the Julian calendar year (365 day calendar). Is there a global or local variable for the day of year in MPASO like there is in POP2 (CESM2). In CESM2, i think you can simply use iday_of_year - but the only potential integer definition of the day of year i'm seeing might be in the orbit subroutine of mpas_ocn_vel_tidal_potential.f90 .

Thanks for any help,
Jacob

**Comment by philipwjones** [comment]
You can get DOY using mpas_get_clock_time to first retrieve a time in a time type and then mpas_get_time with that time and the optional arg DOY to extract a doy. These are in the mpas framework timekeeping module. Interfaces are:

`   subroutine mpas_get_time(curr_time, YYYY, MM, DD, DoY, H, M, S, S_n, S_d, dateTimeString, ierr)

      type (MPAS_Time_type), intent(in) :: curr_time
      integer, intent(out), optional :: YYYY
      integer, intent(out), optional :: MM
      integer, intent(out), optional :: DD
      integer, intent(out), optional :: DoY
      integer, intent(out), optional :: H
      integer, intent(out), optional :: M
      integer, intent(out), optional :: S
      integer, intent(out), optional :: S_n
      integer, intent(out), optional :: S_d
      character (len=StrKIND), intent(out), optional :: dateTimeString
      integer, intent(out), optional :: ierr
`

And

`   type (MPAS_Time_type) function mpas_get_clock_time(clock, whichTime, ierr)

      implicit none

      type (MPAS_Clock_type), intent(in) :: clock
      integer, intent(in) :: whichTime
      integer, intent(out), optional :: ierr
`

**Reply by jacob-stu-allen** [reply]
Very helpful - thank you, Phil!

**Comment by jacob-stu-allen** [comment]
Hello again Phil, I'm getting some errors when attempting to pull the day of year from these two subroutines. Here's what I'm attempting in mpas_ocn_surface_bulk_forcing.f90. I've only included the code relevant to timekeeping. 


```
use mpas_timekeeping
use mpas_timer
...

implicit none
...

subroutine wind_bolster_user(
                               forcingPool, sfcStress, &
                               sfcStressMag,               &
                               clock,                             &
                               err) ! This subroutine is mostly a copy of 
... 
      type (MPAS_Clock_type), intent(in) :: clock
      type (MPAS_Time_type) :: now
... 
      now = mpas_get_clock_time(clock, MPAS_NOW, ierr=err)
      call mpas_get_time( curr_time=now, DoY=refDay, ierr=ierr )

!i'm then using refDay to index a two-dimensional array.

end subroutine wind_bolster_user

```
[logs.zip](https://github.com/E3SM-Project/E3SM/files/13743011/logs.zip)

The E3SM log returns a "Program received signal SIGSEGV: Segmentation fault - invalid memory reference," citing issues in ESMF_ClockMod.f90:518. I've attached my log files as well.

**Reply by philipwjones** [reply]
The error message and location seems to indicate the clock being passed into wind_bolster_user has not been created or initialized correctly.
